### ISSUE TITLE
Resolve load order ambiguities for active plugins

### ISSUE THREAD

**Issue by Ortham** [Ortham/libloadorder#86]
In testing Starfield's sTestFile behaviour, I've found that the load order of plugins activated by sTestFile entries is according to their file modification timestamps, with plugins loading in ascending timestamp order. I'm not sure what happens when two plugins have the same timestamp, I still need to test that.

This can be used to resolve the ambiguity of load order positions for plugins activated using sTestFile, and is probably the same for Skyrim, Skyrim SE, Skyrim VR, Fallout 4 and Fallout 4 VR too. However, for the Skyrim games I need to double-check how sTestFile plugins load in relation to plugins.txt plugins: in the past I've seen the former load after the latter, but that might have just been due to timestamps rather than sorting them in separate blocks, so it might be possible to interleave plugins activated by the two sources.

This doesn't solve the problem of deciding in what order to list plugins that are not active and have no set load order (e.g. because they're not in plugins.txt or the CCC file or because those files are being ignored). That's a different problem, and I don't think there's a good solution to that - at best we could choose to display them in the same order as the game's load order management interface (e.g. launcher or Steam Workshop integration) does, but that's still a fairly arbitrary choice given that order is as good as any other to the game.

To do:

- [x] Test if sTestFile entries are loaded in ascending file modification timestamp order.
- [x] Determine what order sTestFile entries are loaded in if they have the same file modification timestamp.
- [x] Determine if, depending on plugin timestamps, sTestFile plugins can be interleaved with plugins.txt plugins in the load order.
- [x] Update libloadorder to sort sTestFile entries in the same way as the games do.
- [x] Update libloadorder to resolve the ambiguity of equal-timestamp plugins for timestamped-based games in the same way as those games do.

**Comment by Ortham** [comment]
Here are the results of my testing across all relevant games (WIP):


| Game | sTestFile order | sTestFile fallback order | Can interleave plugins.txt and sTestFile plugins? |
|-------|----------------|------------------------|-------------------------------------------------|
| Oblivion | Ascending file modification timestamp | | Yes, plugins.txt order doesn't matter
| Skyrim | Unclear: changing timestamps, filenames, sTestFile indexes and order of sTestFile entries all had no effect on load order. | | | 
| Skyrim SE | Ascending file modification timestamp | Descending filename order | No, with one plugin activated by sTestFile and another by plugins.txt, their order did not change whether one had the higher timestamp or the other had it.
| Fallout 3 | Ascending file modification timestamp | | Yes, plugins.txt order doesn't matter
| Fallout New Vegas | Ascending file modification timestamp | Descending filename order | Yes, plugins.txt order doesn't matter
| Fallout 4 | | | No, plugins.txt is ignored when sTestFile plugins are defined
| Starfield | Ascending file modification timestamp | Ascending filename order | No, plugins.txt is ignored when sTestFile plugins are defined

**Comment by Ortham** [comment]
In the `reduce-ambiguity` branch I've made changes with the following effects:

- For the timestamp-based games:
  - Plugins with equal timestamps are now sorted in descending filename order, they were previously sorted in ascending filename order.
  - `is_ambiguous` now always returns `Ok(false)`.
- For Skyrim and asterisk-based games:
  - Plugins that are installed but not listed in loadorder.txt (or plugins.txt, if loadorder.txt does not exist) are sorted in ascending file modification timestamp order, with files with equal timestamps being sorted in descending filename order. If a plugin's timestamp can't be read, it is sorted earlier than those that can be read. The sorted files are appended to the end of the load order (with adjustments for the usual rules like masters loading after non-masters).
  - This change affects unlisted inactive plugins as well as implicitly active plugins, but retains consistency between the two and the behaviour is now more consistent with xEdit and Wrye Bash (which sort such plugins by file modification timestamp).
- For asterisk-based games:
  - `is_ambiguous()` now treats the positions of all implicitly active plugins as unambiguous, rather than just early-loading plugins.

**Comment by Ortham** [comment]
Done as of 28b3ad1a7b4bd0a0c7afce59ec51fa691648483f. I've left the behaviour for Skyrim unchanged.

**Comment by Ortham** [comment]
Reopening this because Utumno asked if the fallback filename comparison was really case-sensitive, and I tested using Oblivion and found it wasn't (details below). I haven't yet tested the behaviour of the other games.

---

Tested using Oblivion.
Created two copies of DLCMehrunesRazor.esp, called A.esp and B.esp.
Change the name of 005FD6 (Bladeturn Hood):

- A.esp: Bladeturn H00d
- B.esp: Bladeturn H01d

List both in plugins.txt (A.esp then B.esp, the order isn't used though).
Set the modification timestamps A.esp to equal B.esp's.

Loaded the game, checked which indexes the two plugins loaded at:

03 -> H01d (B.esp)
04 -> H00d (A.esp)

Renamed A.esp to a.esp:

03 -> H01d (B.esp)
04 -> H00d (a.esp)

So it looks like the comparison is case-insensitive, but how?

Renamed B.esp to [.esp (lexicographically A < [ < a):

03 -> H01d ([.esp)
04 -> H00d (A.esp)

For completeness, renamed A.esp to a.esp:

03 -> H01d ([.esp)
04 -> H00d (a.esp)

So it looks like the names are uppercased before comparison.

**Comment by Ortham** [comment]
I've tested the case-(in)sensitivity of the fallback filename comparison for Oblivion, Skyrim SE, Fallout 3, Fallout NV, Fallout 4 and Starfield. In all cases, the order is descending uppercased filenames.

This means that Starfield's behaviour has either changed, or I made a mistake in my original testing, despite double-checking it. 

In this testing I've also found that for sTestFile entries specifically, Starfield loads them **in ascending order of their sTestFile indexes**, not ordering by timestamps or filenames. That's also new, and unique to Starfield. It's also worth noting that the sTestFile entries get written to plugins.txt before other installed plugins (which are listed in timestamp order, falling back to descending uppercased order), even if the sTestFile entry plugins have timestamps more recent than some of those other plugins.

For future reference, the Starfield version I tested here was v1.16.236.0.

<details>
<summary>My testing notes</summary>

Fallout 3:

plugins.txt:

A.esp
[.esp

Fallout.ini:

STestFile4=[.esp
STestFile3=A.esp

06 -> The Knowby A01 ([.esp)
07 -> The Knowby A00 (A.esp)

---

Fallout New Vegas:

plugins.txt:

A.esm
[.esm

0a -> Gauss1Rifle ([.esm)
0b -> Gauss0Rifle (A.esm)

---

Skyrim SE

Skyrim.ini:

sTestFile4=A.esp
sTestFile5=[.esp

07 -> Armored1Mage Robe With Enchantment ([.esp)
08 -> Armored0Mage Robe With Enchantment (A.esp)

---

Fallout 4

Fallout3Custom.ini:

sTestFile1=[.esp
sTestFile2=A.esp

fe000 -> 1ludgeoner ([.esm)
fe001 -> 0ludgeoner (A.esm)

Then with Fallout3Custom.ini:

sTestFile1=A.esp
sTestFile2=[.esp

fe000 -> 1ludgeoner ([.esm)
fe001 -> 0ludgeoner (A.esm)

---

Starfield

StarfieldCustom.ini:

sTestFile1=[.esm
sTestFile2=A.esm

04 -> AUG A[ ([.esm)
05 -> AUG AA (A.esm)

04 -> AUG A[ ([.esm)
05 -> AUG AA (a.esm)

This doesn't match my previous testing, but I do also see 

*[.esm
*a.esm

appended to plugins.txt.

I swapped the order to:

sTestFile1=A.esm
sTestFile2=[.esm

and removed the entries in plugins.txt, and on launching Starfield plugins.txt became:

*a.esm
*[.esm

I also tried:

sTestFile2=A.esm
sTestFile1=[.esm

and it gave plugins.txt:

*[.esm
*a.esm

I started the game with plugins.txt containing:

*kgcdoom.esm

and StarfieldCustom.ini:

sTestFile2=A.esm
sTestFile1=[.esm

on start, plugins.txt was rewritten to have:

kgcdoom.esm
*[.esm
*a.esm

(ignoring other plugins). So it looks like sTestFile entries still cause plugins.txt's existing content to be ignored, but sTestFile entries are loaded in ascending order of their indexes, rather than by timestamp or filename.

If I empty plugins.txt and remove the sTestFile entries from the ini, starting Starfield writes plugins.txt to have:

[.esm
a.esm

ignoring other plugins.

If I empty plugins.txt again and set [.esm's timestamp to be 1 second after a.esm's, Starfield writes:

a.esm
[.esm

So the timestamp is used for listing non-sTestFile plugins.

If I rename a.esm to A.esm and set the timestamps equal again, Starfield writes:

[.esm
A.esm

If I rename [.esm to B.esm and set the timestamps equal again, Starfield writes:

B.esm
A.esm

So this indicates that the fallback order for equal timestamps is descending uppercased order, same as for the other games.

I set B.esm's timestamp to be 1 second after A.esm's, and set StarfieldCustom.ini:

sTestFile2=A.esm
sTestFile1=B.esm

and cleared plugins.txt. On launch, Starfield wrote plugins.txt to have:

*B.esm
*A.esm

So it looks like sTestFile index order is used in favour of timestamp order.

It's also worth noting that the sTestFile entries get written to plugins.txt before other installed plugins (which are listed in timestamp order, falling back to descending uppercased order), even if the sTestFile entry plugins have timestamps more recent than some of those other plugins.
</details>
